src/nn/transformer.rs

```diff
@@ -0,0 +1,1097 @@
+use crate::prelude::*;
+use rand::Rng;
+
+/// A multi-head attention layer.
+///
+/// # Generics
+/// - `M` The embedding size of token vectors from decoder.
+/// - `N` The embedding size of token vectors from encoder.
+/// - `K` The size of the keys in self attention.
+/// - `V` The size of the values.
+/// - `H` The number of attention heads.
+///
+/// # Examples
+/// `MultiHeadAttention<8, 10, 10, 10, 2>` is an attention layer with 2 heads and 10 token, key and value dims.
+/// TODO: Doctests fail for some reason
+#[derive(Debug, Clone, Default)]
+pub struct MultiHeadAttention<
+    const M: usize,
+    const N: usize,
+    const K: usize,
+    const V: usize,
+    const H: usize,
+> {
+    w_q: Linear<N, K>,
+    w_k: Linear<M, K>,
+    w_v: Linear<M, V>,
+    w_o: Linear<V, M>,
+}
+
+impl<const M: usize, const N: usize, const K: usize, const V: usize, const H: usize> ResetParams
+    for MultiHeadAttention<M, N, K, V, H>
+{
+    fn reset_params<R: Rng>(&mut self, rng: &mut R) {
+        self.w_q.reset_params(rng);
+        self.w_k.reset_params(rng);
+        self.w_v.reset_params(rng);
+        self.w_o.reset_params(rng);
+    }
+}
+
+impl<const M: usize, const N: usize, const K: usize, const V: usize, const H: usize>
+    CanUpdateWithGradients for MultiHeadAttention<M, N, K, V, H>
+{
+    fn update<G: GradientProvider>(&mut self, grads: &mut G) {
+        self.w_q.update(grads);
+        self.w_k.update(grads);
+        self.w_v.update(grads);
+        self.w_o.update(grads);
+    }
+}
+
+/// Normal self attention (where same tensors are used for keys, queries and values)
+impl<const M: usize, const K: usize, const V: usize, const S: usize, const H: usize>
+    Module<Tensor2D<S, M>> for MultiHeadAttention<M, M, K, V, H>
```

Comment:
We'll need to add a `TAPE: Tape` generic arg here and add to Input/Output. Also will need to do some tape manipulation in the actual forward functions to make sure all the ops are on the tape

Reply:
When passing tensors to matmul, why does the RHS have a NoneTape? Don't the gradients flow to both operators?

Reply:
> When passing tensors to matmul, why does the RHS have a NoneTape? Don't the gradients flow to both operators?

i think because the gradients of rhs are stored in the tape of lhs

Reply:
Exactly yep, just a convention that I went with - rhs will never own the tape in all the binary ops, lhs will either have an OwnedTape or NoneTape. the tape is extracted from lhs and used to add ops to for both operators.

For tracing, none of the stored parameters in a module will ever actually own the tape, the tape is always inserted into the input into the network and moved to intermediate results. Just so happens that all the nn operations so far have parameters as rhs arguments.